airflow-core/src/airflow/models/deadline.py

```diff
@@ -98,6 +98,61 @@ def _determine_resource() -> tuple[str, str]:
             f"{self.deadline_time} or run: {self.callback}({callback_kwargs})"
         )
 
+    @classmethod
+    @provide_session
+    def remove_deadlines(cls, *, session: Session, conditions: dict[Column, Any]) -> int:
+        """
+        Remove deadlines from the table which match the provided conditions and return the number removed.
+
+        NOTE: This should only be used to remove deadlines which are associated with
+            successful dagruns. If the deadline was missed, move it to the `missed_deadlines`
+            table after executing the callback.
+        TODO:  Create the missed_deadlines table (Ramit)
+
+        :param conditions: Dictionary of conditions to evaluate against.
+        :param session: Session to use.
+        """
+        from airflow.models import DagRun  # Avoids circular import
+
+        # Assemble the filter conditions.
+        filter_conditions = [column == value for column, value in conditions.items()]
+        if not filter_conditions:
+            return 0
+
+        try:
+            # Get deadlines which match the provided conditions and their associated DagRuns.
+            deadline_dagrun_pairs = (
+                session.query(Deadline, DagRun).join(DagRun).filter(and_(*filter_conditions)).all()
+            )
+        except SQLAlchemyError as e:
+            invalid_column = next(iter(conditions.keys()))  # Get the first key that caused the error.
+            message = f"Invalid column '{invalid_column}' specified in conditions while resolving deadlines. Rolling back changes."
+            logger.exception(message)
+            session.rollback()
```

Comment:
What's being rolled back here? Wasn't there just a query not a write/transaction?